sunpy/coordinates/frames.py

```diff
@@ -605,6 +605,70 @@ def make_3d(self):
                                                           lat=lat,
                                                           distance=d))
 
+    @u.quantity_input
+    def is_visible(self, *, tolerance: u.m = 1*u.m):
+        """
+        Returns whether the coordinate is on the visible side of the Sun.
+
+        A coordinate is visible if it can been seen from the observer (the ``observer``
+        frame attribute) assuming that the Sun is an opaque sphere with a fixed radius
+        (the ``rsun`` frame attribute).  The visible side of the Sun is always smaller
+        than a full hemisphere.
+
+        Parameters
+        ----------
+        tolerance : `~astropy.units.Quantity`
+            The depth below the surface of the Sun that should be treated as
+            transparent.
+
+        Notes
+        -----
+        If the coordinate is 2D, it is immediately assumed to be visible.
+
+        Due to the limitations of numerical precision, a coordinate that is intended to
+        be on the solar surface might be calculated to be slightly below the surface.
+        The ``tolerance`` parameter accommodates such situations.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> import astropy.units as u
+        >>> from astropy.coordinates import SkyCoord
+        >>> from sunpy.coordinates import HeliographicStonyhurst, Helioprojective
+
+        >>> hpc_frame = Helioprojective(observer='earth', obstime='2023-08-03')
+
+        >>> in_front = SkyCoord(0*u.arcsec, 0*u.arcsec, 0.5*u.AU, frame=hpc_frame)
+        >>> print(in_front.is_visible())
+        True
+
+        >>> behind = SkyCoord(0*u.arcsec, 0*u.arcsec, 1.5*u.AU, frame=hpc_frame)
+        >>> print(behind.is_visible())
+        False
+
+        >>> hgs_array = SkyCoord(np.arange(-180, 180, 60)*u.deg, [0]*6*u.deg,
+        ...                      frame='heliographic_stonyhurst', obstime='2023-08-03')
+        >>> print(hgs_array)
+        <SkyCoord (HeliographicStonyhurst: obstime=2023-08-03T00:00:00.000, rsun=695700.0 km): (lon, lat) in deg
+            [(-180., 0.), (-120., 0.), ( -60., 0.), (   0., 0.), (  60., 0.),
+             ( 120., 0.)]>
+        >>> print(hgs_array.transform_to(hpc_frame).is_visible())
+        [False False  True  True  True False]
+        """
+        # If the coordinate is 2D, it must be visible
+        if self._is_2d:
+            return np.ones_like(self.data, dtype=bool)
+
+        data = self.cartesian
+        data_to_sun = self.observer.radius * CartesianRepresentation(1, 0, 0) - data
+
+        is_behind = data.x < 0
+        is_beyond_limb = np.sqrt(1 - (data.x / data.norm()) ** 2) > self.rsun / self.observer.radius
```

Comment:
```suggestion
        is_beyond_limb = np.sqrt(self.Tx**2 + self.Ty**2) > np.arcsin(self.rsun / self.observer.radius)
```
I got a bit confused with the logic, and I think my suggestion is a bit clearer (making it epxlicit the angle of the coordinate is being compared to the angle of the solar limb)?

Reply:
Your suggestion is formally inaccurate because it relies on the small-angle approximation to combine Tx and Ty like that to get the radial angle.

Reply:
Could you add a comment on what `data.x` is instead then? I think it would be good to make it easier for someone else to understand (and check) the maths going on here.

Reply:
I added a clarification for `data.x`, and I also changed the expression to `np.sqrt(data.y **2 + data.z **2) / data.norm()` since that's likely easier for future readers.